packages/connect/src/protocol/universal-handler-client.ts

```diff
@@ -38,23 +39,62 @@ export function createUniversalHandlerClient(
         Code.Unimplemented
       );
     }
-    const uServerRes = await handler({
-      body: uClientReq.body,
-      httpVersion: "2.0",
-      method: uClientReq.method,
-      url: reqUrl,
-      header: uClientReq.header,
-      signal: uClientReq.signal ?? new AbortController().signal,
-    });
+    const reqSignal = uClientReq.signal ?? new AbortController().signal;
+    const uServerRes = await raceSignal(
+      reqSignal,
+      handler({
+        body: uClientReq.body,
+        httpVersion: "2.0",
+        method: uClientReq.method,
+        url: reqUrl,
+        header: uClientReq.header,
+        signal: reqSignal,
+      })
+    );
     let body = uServerRes.body ?? new Uint8Array();
     if (body instanceof Uint8Array) {
       body = createAsyncIterable([body]);
     }
     return {
-      body: body,
+      body: pipe(body, (iterable) => {
+        return {
+          [Symbol.asyncIterator]() {
+            const it = iterable[Symbol.asyncIterator]();
+            const w: AsyncIterator<Uint8Array> = {
+              next() {
+                return raceSignal(reqSignal, it.next());
+              },
+            };
+            if (it.throw !== undefined) {
+              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- can't handle mutated object sensibly
+              w.throw = (e: unknown) => it.throw!(e);
+            }
+            if (it.return !== undefined) {
+              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion,@typescript-eslint/no-explicit-any -- can't handle mutated object sensibly
+              w.return = (value?: any) => it.return!(value);
+            }
+            return w;
+          },
+        };
+      }),
       header: new Headers(uServerRes.header),
       status: uServerRes.status,
       trailer: new Headers(uServerRes.trailer),
     };
   };
 }
+
+function raceSignal<T>(signal: AbortSignal, promise: Promise<T>): Promise<T> {
+  let cleanup: (() => void) | undefined;
+  const signalPromise = new Promise<never>((_, reject) => {
+    if (signal.aborted) {
+      reject(getAbortSignalReason(signal));
```

Comment:
Could you just call `onAbort` here? Or do you intentionally want to keep them separate?

Reply:
Good call, updated.